dali/imgcodec/decoders/decoder_parallel_impl.h

```diff
@@ -30,108 +33,122 @@ namespace imgcodec {
  */
 class DLL_PUBLIC BatchParallelDecoderImpl : public ImageDecoderImpl {
  public:
-  BatchParallelDecoderImpl(int device_id, ThreadPool *tp)
-  : ImageDecoderImpl(device_id, tp) {}
+  explicit BatchParallelDecoderImpl(int device_id)
+  : ImageDecoderImpl(device_id) {}
 
   using ImageDecoderImpl::CanDecode;
-  std::vector<bool> CanDecode(cspan<ImageSource *> in,
+  std::vector<bool> CanDecode(DecodeContext ctx,
+                              cspan<ImageSource *> in,
                               DecodeParams opts,
                               cspan<ROI> rois) override {
     assert(rois.empty() || rois.size() == in.size());
     std::vector<bool> ret(in.size());
     ROI no_roi;
     for (int i = 0; i < in.size(); i++) {
-      tp_->AddWork([&, i](int) {
-        ret[i] = CanDecode(in[i], opts, rois.empty() ? no_roi : rois[i]);
+      ctx.tp->AddWork([&, ctx, i](int) {
+        ret[i] = CanDecode(ctx, in[i], opts, rois.empty() ? no_roi : rois[i]);
       });
     }
-    tp_->RunAll();
+    ctx.tp->RunAll();
     return ret;
   }
 
   using ImageDecoderImpl::Decode;
-  std::vector<DecodeResult> Decode(span<SampleView<CPUBackend>> out,
+  std::vector<DecodeResult> Decode(DecodeContext ctx,
+                                   span<SampleView<CPUBackend>> out,
                                    cspan<ImageSource *> in,
                                    DecodeParams opts,
                                    cspan<ROI> rois) override {
-    return ParallelDecodeImpl(out, in, opts, rois);
+    auto future = ScheduleDecode(std::move(ctx), out, in, opts, rois);
+    auto res = future.get_all(true);
```

Comment:
is there a point of allowing `get_all(false)` at all?

Reply:
Removed after checking that get_one is efficient enough.